velox/functions/lib/DateTimeFormatter.h

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+#include <vector>
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/Timestamp.h"
+
+namespace facebook::velox::functions {
+
+static constexpr size_t kStructReserveSize{8};
+static constexpr size_t kLiteralReserveSize{64};
+
+enum class DateTimeFormatSpecifier : uint8_t {
+  // Era, e.g: "AD"
+  ERA = 0,
+
+  // Century of era (>=0), e.g: 20
+  CENTURY_OF_ERA = 1,
+
+  // Year of era (>=0), e.g: 1996
+  YEAR_OF_ERA = 2,
+
+  // Year of era (>=0), e.g: 1996
+  WEEK_YEAR = 3,
+
+  // Week of weekyear, e.g: 27
+  WEEK_OF_WEEK_YEAR = 4,
+
+  // Day of week, 0 ~ 6 with 0 representing Sunday
+  DAY_OF_WEEK_0_BASED = 5,
+
+  // Day of week, 1 ~ 7
+  DAY_OF_WEEK_1_BASED = 6,
+
+  // Day of week, e.g: "Tuesday" or "Tue", depending on number of times provided
+  // formatter character repeats
+  DAY_OF_WEEK_TEXT = 7,
+
+  // Year, e.g: 1996
+  YEAR = 8,
+
+  // Day of year, e.g: 189
+  DAY_OF_YEAR = 9,
+
+  // Month of year, e.g: 07, or 7 depending on number of times provided
+  // formatter character repeats
+  MONTH_OF_YEAR = 10,
+
+  // Month of year, e.g. Dec, December depending on number of times provided
+  // formatter character repeats
+  MONTH_OF_YEAR_TEXT = 11,
+
+  // Day of month, e.g: 10, 01, 001, with/without padding 0s depending on number
+  // of times provided formatter character repeats
+  DAY_OF_MONTH = 12,
+
+  // Halfday of day, e.g: "PM"
+  HALFDAY_OF_DAY = 13,
+
+  // Hour of halfday (0~11)
+  HOUR_OF_HALFDAY = 14,
+
+  // Clockhour of halfday (1~12)
+  CLOCK_HOUR_OF_HALFDAY = 15,
```

Comment:
Not 0- or 1-based, like "day of week" above?

Reply:
I think it is self explanatory, right? You don't have 0 in a clock that's why it's called clock hour.

Reply:
That is not straightforward and obvious only when you work with it. :)
But not a big deal, just pointed out at some inconsistency in naming.